compiler/src/dotty/tools/dotc/transform/Inlining.scala

```diff
@@ -88,6 +93,11 @@ class Inlining extends MacroTransform, IdentityDenotTransformer {
 
             flatTree(trees2)
           else super.transform(tree)
+        case Apply(fun, args) if fun.symbol.name == nme.unapply && fun.symbol.is(ConstructorProxy) =>
```

Comment:
Should this transformation happen in the `Typer` just after we've created the `unapply` object instead?